File: viewpump/src/main/java/io/github/inflationx/viewpump/InterceptorChain.java
package io.github.inflationx.viewpump;

import android.support.annotation.NonNull;

import java.util.List;

/**
 * A concrete interceptor chain that carries the entire interceptor chain.
 */
class InterceptorChain implements Interceptor.Chain {
    private final List<Interceptor> interceptors;
    private final int index;
    private final InflateRequest request;

    InterceptorChain(@NonNull List<Interceptor> interceptors, int index, @NonNull InflateRequest request) {
        this.interceptors = interceptors;
        this.index = index;
        this.request = request;
    }

    @NonNull
    @Override
    public InflateRequest request() {
        return request;
    }

    @NonNull
    @Override
    public InflateResult proceed(@NonNull InflateRequest request) {
        if (index >= interceptors.size()) {
            throw new AssertionError("no interceptors added to the chain");
        }

        // Call the next interceptor in the chain.
        InterceptorChain next = new InterceptorChain(interceptors, index + 1, request);
        Interceptor interceptor = interceptors.get(index);
        InflateResult result = interceptor.intercept(next);

        // Confirm that the intercepted response isn't null.
        if (result == null) {
            throw new NullPointerException("interceptor " + interceptor + " returned null");
        }

        return result;
    }
}

File: viewpump/src/main/java/io/github/inflationx/viewpump/internal/-InterceptorChain.kt
@file:JvmName("-InterceptorChain")
package io.github.inflationx.viewpump.internal

import io.github.inflationx.viewpump.InflateRequest
import io.github.inflationx.viewpump.InflateResult
import io.github.inflationx.viewpump.Interceptor
import io.github.inflationx.viewpump.Interceptor.Chain

/**
 * A concrete interceptor chain that carries the entire interceptor chain.
 */
@Suppress("ClassName")
internal class `-InterceptorChain`(private val interceptors: List<Interceptor>, private val index: Int,
    private val request: InflateRequest) : Chain {

  override fun request(): InflateRequest {
    return request
  }

  override fun proceed(request: InflateRequest): InflateResult {
    if (index >= interceptors.size) {
      throw AssertionError("no interceptors added to the chain")
    }

    // Call the next interceptor in the chain.
    val next = `-InterceptorChain`(interceptors, index + 1,
        request)
    val interceptor = interceptors[index]

    return interceptor.intercept(next)
  }
}
